node/core/provisioner/src/lib.rs

```diff
@@ -466,18 +469,42 @@ async fn select_candidates(
 	}
 
 	// keep only one candidate with validation code.
-	let mut with_validation_code = false;
-	candidates.retain(|c| {
-		if c.candidate.commitments.new_validation_code.is_some() {
-			if with_validation_code {
-				return false
-			}
+	let mut with_validation_code = None;
+	let mut cumulative_size = 0usize;
+	let mut out = Vec::new();
+
+	// Drain the candidates list to extract a suitable list to propose
+	while let Some(c) = candidates.pop() {
+		// If candidate exceeds max candidate size, we needn't include them
+		let encoded = c.encode();
+		if encoded.len() > MAX_SINGLE_CANDIDATE_SIZE {
+			continue
+		}
 
-			with_validation_code = true;
+		// If candidate leads to an overlength block, we needn't include them
+		cumulative_size += encoded.len();
+		if cumulative_size > MAX_CANDIDATES_SIZE {
+			break
 		}
 
-		true
-	});
+		// If candidate contains new validation code, we update cumulative_size
+		if c.candidate
+			.commitments
+			.new_validation_code
+			.as_ref()
+			.map(|_| {
+				cumulative_size.saturating_sub(
+					with_validation_code
+						.replace(c.clone())
```

Comment:
If our candidate has `new_validation_code`, we want to replace the old `with_validation_code` via `fn replace`, but we must also remove the old candidate's size from `cumulative_size` in order to ensure we don't overcount the block length.

If `with_validation_code` is `None`, then we can `unwrap_or_default()` to 0.